packages/vite/src/node/shortcuts.ts

```diff
@@ -0,0 +1,134 @@
+import type { HmrOptions } from './server/hmr'
+import { isDefined } from './utils'
+
+export type BindShortcutsOptions = {
+  /**
+   * Print a one line hint to the terminal.
+   */
+  print?: boolean
+  customShortcuts?: (CLIShortcut | undefined | null)[]
+}
+
+export type CLIShortcut = {
+  key: string
+  description: string
+  action(server: ViteDevServer): void | Promise<void>
+}
+
+export function bindShortcuts(
+  server: ViteDevServer,
+  opts: BindShortcutsOptions,
+): void {
+  if (!server.httpServer) return
+
+  if (opts.print) {
+    server.config.logger.info(
+      colors.dim(colors.green('  ➜')) +
+        colors.dim('  press ') +
+        colors.bold('h') +
+        colors.dim(' to show help'),
+    )
+  }
+
+  const shortcuts = (opts.customShortcuts ?? [])
+    .filter(isDefined)
+    .concat(BASE_SHORTCUTS)
+
+  let actionRunning = false
+
+  const onInput = async (input: string) => {
+    // ctrl+c or ctrl+d
+    if (input === '\x03' || input === '\x04') {
+      process.emit('SIGTERM')
+      return
+    }
+
+    if (actionRunning) return
+
+    if (input === 'h') {
+      server.config.logger.info(
+        shortcuts
+          .map(
+            (shortcut) =>
+              colors.dim('  press ') +
+              colors.bold(shortcut.key) +
+              colors.dim(` to ${shortcut.description}`),
+          )
+          .join('\n'),
+      )
+    }
+
+    const shortcut = shortcuts.find((shortcut) => shortcut.key === input)
+    if (!shortcut) return
+
+    actionRunning = true
+    await shortcut.action(server)
+    actionRunning = false
+  }
+
+  if (process.stdin.isTTY) {
+    process.stdin.setRawMode(true)
+  }
+
+  process.stdin.on('data', onInput).setEncoding('utf8').resume()
+
+  server.httpServer.on('close', () => {
+    process.stdin.off('data', onInput).pause()
+  })
+}
+
+let initialHmrOptions: HmrOptions | boolean
+
+const BASE_SHORTCUTS: CLIShortcut[] = [
+  {
+    key: 'r',
+    description: 'restart the server',
+    async action(server) {
+      await server.restart()
+    },
+  },
+  {
+    key: 'o',
+    description: 'open in browser',
+    action(server) {
+      const url = server.resolvedUrls?.local[0]
+
+      if (!url) {
+        server.config.logger.warn('No URL available to open in browser')
+        return
+      }
+
+      openBrowser(url, true, server.config.logger)
+    },
+  },
+  {
+    key: 'm',
+    description: 'toggle hmr on/off',
+    action({ config }: ViteDevServer): void {
+      initialHmrOptions ??= config.server.hmr ?? true
+      /**
+       * Mutating the server config works because Vite reads from
+       * it on every file change, instead of caching its value.
+       */
+      config.server.hmr =
+        config.server.hmr === false
+          ? initialHmrOptions === false
+            ? true
+            : initialHmrOptions
+          : false
+      config.logger.info(
+        colors.cyan(`hmr ${config.server.hmr ? `enabled` : `disabled`}`),
+      )
+    },
+  },
```

Comment:
Some context from the previous PR: https://github.com/vitejs/vite/pull/9673#issuecomment-1215213681

> @patak-dev said: I also have doubts about `h`. When would this option be used?
> 
> @aleclarson said: I added this one. It's useful when you're step debugging a client-side issue and hot-reloading would cause you to lose your state. If we don't have this option, you have to stop the dev server instead. I'd be fine with opening a separate PR for this option in the future, though.
>
> @patak-dev said: How does this debugging setup work? Is the idea that you would like to do changes to your files and they are big enough that you want to have intermediate save points while debugging purely in the client?
> I'm ok with this, ya, it looks useful. Maybe we should add an explanation of when this is useful in the docs too. Let's see what others think about the option.

Reply:
Thanks for the reference, I still think we should discuss it in a separate PR

Reply:
Done. Now the list seems very small. I'm wondering if I should add custom shortcut from vite config so that's available for people not using the JS API?